File: 392/Main.java
import java.io.*;
import java.util.*;

public class Main {
   final static String[] x = {
      "x^8",
      "x^7",
      "x^6",
      "x^5",
      "x^4",
      "x^3",
      "x^2",
      "x",
      "",
   };
   public static void main(String args[]) throws IOException {
      String line;
      StringBuilder answer = new StringBuilder();
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((line = in.readLine()) != null) {
         StringTokenizer tokenizer = new StringTokenizer(line);
         int[] n = new int[9];
         boolean[] negative = new boolean[9];
         for (int i = 0; i < n.length; i++) {
            n[i] = Integer.parseInt(tokenizer.nextToken());
            negative[i] = (n[i] < 0);
            if (negative[i]) n[i] = -n[i];
         }
         int first = 0;
         while (first < (n.length - 1) && n[first] == 0) ++first;
         answer.append((negative[first]) ? "-" : "");
         if (!((n[first] == 1) && (first < (n.length - 1))))
            answer.append(Integer.toString(n[first]));
         answer.append(x[first]);
         for (int i = first + 1; i < n.length; i++) {
            if (n[i] == 0) continue;
            answer.append((negative[i]) ? " - " : " + ");
            if (!((n[i] == 1) && (i < (n.length - 1))))
               answer.append(Integer.toString(n[i]));
            answer.append(x[i]);
         }
         answer.append('\n');
      }
      System.out.print(answer);
   }
}

File: 392/main.c
#include <stdio.h>

const static char x[9][4] = {
   "x^8",
   "x^7",
   "x^6",
   "x^5",
   "x^4",
   "x^3",
   "x^2",
   "x",
   "",
};

int main(int argc, char* argv[])
{
   int i, first;
   int n[9];
   char negative[9];

   while (scanf("%d", &n[0]) == 1) {
      negative[0] = (n[0] < 0) ? 1 : 0;
      if (negative[0]) n[0] = -n[0];
      for (i = 1; i < 9; i++) {
         scanf("%d", &n[i]);
         negative[i] = (n[i] < 0) ? 1 : 0;
         if (negative[i]) n[i] = -n[i];
      }
      first = 0;
      while ((first < 8) && n[first] == 0) ++ first;
      printf("%s", (negative[first] == 1) ? "-" : "");
      if (!((n[first] == 1) && (first < 8))) {
         printf("%d", n[first]);
      }
      printf("%s", x[first]);
      for (i = first + 1; i < 9; i++) {
         if (n[i] == 0) continue;
         printf("%s", (negative[i] == 1) ? " - " : " + ");
         if (!((n[i] == 1) && (i < 8)))
            printf("%d", n[i]);
         printf("%s", x[i]);
      }
      printf("\n");
   }

   return 0;
}
